[CLIPPER-212] Add CLOCK cache eviction policy

Corey-Zumar opened the pull request:
Fixes #212

The cache eviction policy is a slightly-modified version of the clock algorithm. Notably, we don't evict any entries from the cache that have incomplete promises. Additionally, multiple evictions may occur for a single insertion due to the heterogeneity of entry sizes.

Corey-Zumar committed:
Implement CLOCK policy

Corey-Zumar committed:
add cache tests

Corey-Zumar committed:
revert clipper mgr changes

Corey-Zumar committed:
remove quotes

Corey-Zumar committed:
add test assertion

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/607/
Test FAILed.

Corey-Zumar committed:
format code

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/608/
Test PASSed.

Corey-Zumar commented on src/libclipper/include/clipper/task_executor.hpp:
```diff
@@ -24,6 +24,7 @@
 namespace clipper {
 
 const std::string LOGGING_TAG_TASK_EXECUTOR = "TASKEXECUTOR";
+constexpr size_t CACHE_SIZE_ELEMENTS = 2048;
```
Because cache entries are of variable size and the user is unaware of the size of entry metadata, I think this should be a configurable value named `CACHE_SIZE_BYTES` instead. However, making this change complicates cache eviction substantially and will reduce CLOCK performance. Let's discuss @dcrankshaw.

Corey-Zumar replied:
Actually, making this change shouldn't be too bad. Adding the "in progress" label instead.

Corey-Zumar replied:
Done!

dcrankshaw replied:
Now that you changed the cache size to be measured in bytes, should this variable be deleted?

Corey-Zumar committed:
Refactor PredictionCache to evict based on byte size rather than num pages

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/609/
Test FAILed.

Corey-Zumar committed:
Add config flag for cache size

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/610/
Test FAILed.

Corey-Zumar committed:
Format code

Corey-Zumar committed:
Update config tests

Corey-Zumar committed:
format code

Corey-Zumar commented:
@dcrankshaw This is now ready for review.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/611/
Test PASSed.

Corey-Zumar committed:
Update config.hpp

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/612/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/613/
Test FAILed.

Corey-Zumar commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/614/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/615/
Test PASSed.

dcrankshaw commented on src/frontends/src/query_frontend_main.cpp:
```diff
@@ -11,18 +11,21 @@ int main(int argc, char* argv[]) {
     ("redis_port", "Redis port",
-        cxxopts::value<int>()->default_value("6379"))
+        cxxopts::value<int>()->default_value(std::to_string(clipper::DEFAULT_REDIS_PORT)))
     ("threadpool_size", "Number of threads for the task execution threadpool",
```
You can actually delete this command line arg and corresponding config field entirely. It's no longer used. Instead we create a thread in the task execution threadpool for each connected container.

Corey-Zumar replied:
Done!

dcrankshaw commented on src/frontends/src/query_frontend_main.cpp:
```diff
@@ -11,18 +11,21 @@ int main(int argc, char* argv[]) {
-        cxxopts::value<int>()->default_value("4"));
+        cxxopts::value<int>()->default_value(std::to_string(clipper::DEFAULT_TASK_EXECUTION_THREADPOOL_SIZE)))
+    ("prediction_cache_size", "Size of the prediction cache in bytes, excluding cache metadata",
+       cxxopts::value<long>()->default_value(std::to_string(clipper::DEFAULT_PREDICTION_CACHE_SIZE_BYTES)));
```
There should a way to set this value from clipper_admin. Add an optional arg to set this in the old version of clipper_admin and I'll make sure it gets ported to v2.

Corey-Zumar replied:
Done: added a `cache_size` argument to the constructor of `clipper_manager.Clipper`

dcrankshaw commented on src/libclipper/src/task_executor.cpp:
```diff
@@ -59,21 +62,68 @@ void PredictionCache::put(const VersionedModelId &model,
+  size_t entry_size_bytes = value.completed_ ? value.value_.y_hat_.size() : 0;
+  if (entry_size_bytes <= max_size_bytes_) {
+    evict_entries(size_bytes_ + entry_size_bytes - max_size_bytes_);
+    page_buffer_.insert(page_buffer_.begin() + page_buffer_index_, key);
```
What is the page buffer?

Corey-Zumar replied:
The `page_buffer_` is a vector of input hashes and acts as the clock in the cache eviction algorithm. In conjunction with the buffer (clock) itself, `page_buffer_index_` acts as the clock hand. To examine a page buffer entry with key `K`, we query the map `entries_` with `K` as the input hash key. We then determine whether or not the entry has all of its futures completed (via `entry.completed_`) and whether or not it was recently used (via the `entry.used_` boolean flag). If the entry is completed and was not recently used, we remove the key `K` from the page buffer, and the entry itself is removed from the `entries_` map.

dcrankshaw commented on src/libclipper/src/task_executor.cpp:
```diff
@@ -59,21 +62,68 @@ void PredictionCache::put(const VersionedModelId &model,
+    return;
+  }
+  while (space_needed_bytes > 0) {
+    long page_key = page_buffer_[page_buffer_index_];
```
What is the role of the page_buffer and page_buffer_index here?

Corey-Zumar replied:
See the above comment. We use `page_buffer_` and `page_buffer_index_` to evict as many pages as necessary to obtain the required free space.

dcrankshaw reviewed (changes requested):
Overall looks good. I added a comment asking for clarification about how eviction is implemented.

Do you think this will impose a large overhead on workloads with a low cache hit rate? Should we think about adding a cache eviction policy that very cheaply evicts any entries that have already been fetched? We would need to ensure that an entry is only ever fetched from the cache once-per-query.

Corey-Zumar committed:
Remove unused task executor threadpool size config

Corey-Zumar committed:
Add cache size option to clipper mgr

Corey-Zumar committed:
format code

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/620/
Test FAILed.

Corey-Zumar committed:
remove task execution variable

Corey-Zumar committed:
when evicting, stop if we're out of pages

Corey-Zumar commented:
@dcrankshaw Addressed your review comments. Regarding applications with low cache hit rates, the user can avoid eviction overhead by setting the cache size to zero, effectively disabling caching.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/621/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/622/
Test FAILed.

Corey-Zumar committed:
remove task execution test statement

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/623/
Test PASSed.

dcrankshaw reviewed (approved):
LGTM

Corey-Zumar committed:
Fix rebase issue

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/646/
Test FAILed.

Corey-Zumar committed:
fix test

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/647/
Test FAILed.

Corey-Zumar commented:
jenkins test this please

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/648/
Test PASSed.
